roles/maps/tasks/install_frontend.yml

```
@@ -76,6 +76,26 @@
     dest: "{{ maps_serve_path }}/{{ maps_dot_black_satellite_symlink_name }}"
     state: link
 
+- when: maps_terrain_zoom != "none"
+  block:
```

~Reverse the order of these 2 lines I think!~

```suggestion
- block:
  when: maps_terrain_zoom != none
```

PS keep the quotes around "none" if you strongly prefer that (i personally don't ;)

---

Ansible linter specifically told me to set this ordering of lines.

---

Ah maybe you're looking at the warning from before I changed it?

---

And it appears as though the quotes are required for this test to work properly. Indeed, the other cases of `when: ... "none"` in the codebase use quotes.

---

> Ansible linter specifically told me to set this ordering of lines.

Oops, looks like I was fooled by the lack of a `- name:` line on top.

---

> And it appears as though the quotes are required for this test to work properly. Indeed, the other cases of `when: ... "none"` in the codebase use quotes.

Worrisome, if `none` and/or `None` have a special meaning in Ansible, we need to know.

I need to dig deeper.  And if we really have to revert to `null` or something else more standardized, to truly clean up, we can do that.

---

$ ansible-console localhost -c local
    ...
    @xk@localhost (1)[f:5]$ debug msg="{{ None | type_debug }}"
    localhost | SUCCESS => {
        "msg": "NoneType"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ "None" | type_debug }}"
    localhost | SUCCESS => {
        "msg": "str"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ "none" | type_debug }}"
    localhost | SUCCESS => {
        "msg": "str"
    }

"none" is fine

Actually, it looks like `null` isn't even a thing:

    @xk@localhost (1)[f:5]$ debug msg="{{ null | type_debug }}"
    localhost | SUCCESS => {
        "msg": "UndefinedMarker"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ aaa | type_debug }}"
    localhost | SUCCESS => {
        "msg": "UndefinedMarker"
    }

---

> @xk@localhost (1)[f:5]$ debug msg="{{ None | type_debug }}"
> localhost | SUCCESS => {
>     "msg": "NoneType"
> }

Likewise...

```
root@localhost (1)[f:5]$ debug msg="{{ none is none }}"
localhost | SUCCESS => {
    "msg": true
}
```

(Important revelation, if a bit scary as to what we should do about this, i.e. the fact that `none` and `None` are very different from `"none"` and `"None"` !)

---

@xk@localhost (1)[f:5]$ debug msg="{{ none | type_debug }}"
    localhost | SUCCESS => {
        "msg": "NoneType"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ None }}"
    localhost | SUCCESS => {
        "msg": null
    }

`none` == `None`

but "None" != "none" != None

    @xk@localhost (1)[f:5]$ debug msg="{{ 'None' == None }}"
    localhost | SUCCESS => {
        "msg": false
    }

    @xk@localhost (1)[f:5]$ debug msg="{{ 'none' is None }}"
    ... <stacktrace>
    {{ 'none' is 'none' }}
    ...
    localhost | FAILED! => {
        "changed": false,
        "msg": "Task failed: Finalization of task args for 'ansible.builtin.debug' failed: Error while resolving value for 'msg': Syntax error in template: expected token 'name', got 'string'"
    }

for the same reason that Python will print a SyntaxWarning

    >>> 'none' is None
    <python-input-0>:1: SyntaxWarning: "is" with 'str' literal. Did you mean "=="?
    False

---

Also note you can use `truthy` and `falsy` to approximate the `bool()` python function:

    @xk@localhost (1)[f:5]$ debug msg="{{ None is truthy }}"
    localhost | SUCCESS => {
        "msg": false
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ 'none' is truthy }}"
    localhost | SUCCESS => {
        "msg": true
    }

---

I initially thought that `none` was a special thing, but then it seems like setting the new variable to `none` vs `"none"` in local_vars has the same effect. My conclusion was that it's a string, but that `when:` clauses needs strings to be in quotes. But given your results here I'm just confused.

---

I think it's fine to use "none" as the default values as the end result is that it is more user friendly.

It's only dangerous if the value is a None value--then this check returns true:

    - when: maps_terrain_zoom != "none"

    @xk@localhost (1)[f:5]$ debug msg="{{ none != 'none' }}"
    localhost | SUCCESS => {
        "msg": true
    }

Using None as the default value may appear safer...:

    - when: maps_terrain_zoom

but in practice it is only a problem if people use none instead of "none"

    maps_terrain_zoom: none

...though you could specifically guard against this condition:

    @xk@localhost (1)[f:5]$ debug msg="{{ None is truthy and None != 'none' }}"
    localhost | SUCCESS => {
        "msg": false
    }

ie.

    - when: maps_terrain_zoom is truthy and maps_terrain_zoom != "none"

---

> problem if people use none instead of "none"

Indeed.  And that could be incredibly serious, when IIAB's implementer/operator community drops-or-adds quotation marks all the time (they cannot be expected to understand YAML & Ansible!)  They want something "Dead Simple" (KISS, etc) after all.  In other words, all these common mistakes will in fact occur within everyday IIAB implementers' local_vars.yml files regularly — as that's just human nature:

- I worry that my/our choice to sanctify the specific word n-o-n-e (which seemed perfect at first!!) might... be a terrible precedent in the end... unintentionally make our entire code base flakier & more brittle ;)

- So maybe we can take a breather, sleep on it etc, and sometime over the coming days see if Santa might find a Far Cleaner Way to solve this fundamental Ansible conundrum (several of us may be able to navigate this minefield of poorly documented Ansible quirks... without too many bruises... _but our implementer community most definitely will not ;)_

🎄

---

> sanctify the specific word

To be more clear, it isn't a problem with the word chosen. The result of `'nothing'` and `'none'` are both strings and will have the above issue.

Without quotes, `none` will resolve to NoneType and `nothing` is an error.

`null` is also [not a reserved word](https://github.com/ansible/ansible/issues/37441)--but its behavior seems to be different between ansible-console and [templates](https://stackoverflow.com/questions/73515934/ansible-variable-null-vs-none) (and maybe this is handled by YAML instead of something that Ansible is specifically doing):

    @xk@localhost (1)[f:5]$ debug msg="{{ null }}"
    localhost | FAILED! => {
        "changed": false,
        "msg": "Task failed: Finalization of task args for 'ansible.builtin.debug' failed: Error while resolving value for 'msg': 'null' is undefined"
    }

But templates act more like this:

    @xk@localhost (1)[f:5]$ debug msg="{{ null | default(None) }}"
    localhost | SUCCESS => {
        "msg": null
    }

But maybe it does that for every undefined instead of raising an error. (Un?)helpfully, Ansible prints None as null.

    @xk@localhost (1)[f:5]$ debug msg="{{ null | default(None) | type_debug }}"
    localhost | SUCCESS => {
        "msg": "NoneType"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ null | type_debug }}"
    localhost | SUCCESS => {
        "msg": "UndefinedMarker"
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ nothing | type_debug }}"
    localhost | SUCCESS => {
        "msg": "UndefinedMarker"
    }

### So what?

I don't think it's possible to handle this "cleanly" from a code perspective if we want pokayoke for the user. Either we use `word` as the default (_whether its a real reserved word like none or the undefined error makes it become None_), or we use `"word"` as the default--the user could still do the opposite and add quotes where we are only checking for the unquoted version or vice-versa.

If we want to handle bad input "correctly" then we need to check both ahead of time. The only way to reduce the amount of code is to handle less bad config input or adjust how things are handled through some global Ansible configuration or macros (if that exists). Otherwise we'll need something like this:

    - when: maps_terrain_zoom is truthy and maps_terrain_zoom != "none"

or perhaps more simply...

    - when: maps_terrain_zoom is not None and maps_terrain_zoom != "none"

...this one has a minor problem in that it evaluates to NoneType instead of false...

    - when: maps_terrain_zoom and maps_terrain_zoom != "none"

...depending on the order:

    @xk@localhost (1)[f:5]$ debug msg="{{ none and false }}"
    localhost | SUCCESS => {
        "msg": null
    }
    @xk@localhost (1)[f:5]$ debug msg="{{ false and none }}"
    localhost | SUCCESS => {
        "msg": false
    }

(But we should probably casefold the string check.)

If we use `null` or `none` as the default value instead of `"none"` then the above would still be what we would want to do if we want to guard against unintended behavior on behalf of the user.